### PR TITLE
tesseract messaging integration test fix #277

### PR DESCRIPTION
**Pull request by MrishoLukamba**

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -354,13 +355,14 @@ jobs:
       #        run: |
       #          cargo test -p polygon-pos-prover -- --nocapture --ignored
 
+      - name: parachain ismp relayer messaging integration tests
+        run: |
+          cargo test -p tesseract-integration-test --lib -- --ignored --test-threads=1
+
```
Comment (seunlanlege) [inline .github/workflows/ci.yml:361]: This should run immediately after the devnet step

File: modules/ismp/pallets/demo/src/lib.rs [hunk 2]
```diff
@@ -113,6 +114,12 @@ pub mod pallet {
 		GetResponse(Vec<Option<Vec<u8>>>),
 	}
 
+	/// Storing values requested from other chains along with their requesting keys
+	/// hash of the key
+	#[pallet::storage]
+	pub type GetResponses<T: Config> =
+		StorageMap<_, Twox128, [u8; 16], Option<Vec<u8>>, ValueQuery>;
+
```
Comment (seunlanlege) [inline modules/ismp/pallets/demo/src/lib.rs:122]: Unused lets remove

File: modules/ismp/pallets/demo/src/lib.rs [hunk 3]
```diff
@@ -349,9 +356,18 @@ impl<T: Config> IsmpModule for IsmpModuleCallback<T> {
...
-			Response::Get(res) => Pallet::<T>::deposit_event(Event::<T>::GetResponse(
-				res.values.into_values().collect(),
-			)),
+			Response::Get(res) => {
+				res.values.iter().for_each(|(k, v)| {
+					if v.is_some() {
+						let hash_key = Twox128::hash(k.as_ref());
+						GetResponses::<T>::insert(hash_key, v)
+					}
+				});
```
Comment (seunlanlege) [inline modules/ismp/pallets/demo/src/lib.rs:365]: Why?

File: tesseract/integration-test/src/lib.rs [hunk 4]
```diff
@@ -112,47 +69,49 @@ async fn relay_get_response_message(
...
 
 		Message::Response(response)
 	};
+	let cloned_chain_a_client = chain_a_client.clone();
+	let event_background = tokio::spawn(
+		async move {
+			let _res = wait_for_event_chain_a(chain_a_client.clone().client.clone()).await;
+		}
+		.boxed(),
+	);
```
Comment (seunlanlege) [inline tesseract/integration-test/src/lib.rs:90]: Once again, this is how we miss events. Get the event using subxt

File: tesseract/integration-test/src/lib.rs [hunk 5]
```diff
@@ -459,19 +454,40 @@ async fn get_request_works() -> Result<(), anyhow::Error> {
...
+			Ok(_block) => {
+				log::info!("chain producing blocks and finalizing");
+				break
+			},
+			Err(_err) => {
+				log::info!("chain not yet producing blocks and finalizing");
+				continue
+			},
+		}
+	}
```
Comment (seunlanlege) [inline tesseract/integration-test/src/lib.rs:474]: Just take 1 and end the stream

File: tesseract/integration-test/src/lib.rs [hunk 6]
```diff
@@ -226,6 +185,55 @@ async fn set_host_params(
...
+				}
+			},
+			Err(err) => {
+				// Log the error and return it as a `Result` error
+				log::error!("Error in finalized block stream: {:?}", err);
+			},
+		}
+	}
+	Ok(())
+}
```
[the remark this answers is not in the document]
Comment (MrishoLukamba) [reply tesseract/integration-test/src/lib.rs:235]: David advised me to fail gracefully as the tx may return invalid response but will be executed due to reorgs so without propagating error, I cannot get blockhash, hence not getting tx event at specific time

File: modules/ismp/pallets/demo/src/lib.rs (hunk 3 above)
Comment (MrishoLukamba) [reply modules/ismp/pallets/demo/src/lib.rs:365]: David advised me to put the returned value in the storage and just read the storage

Comment (seunlanlege) [reply modules/ismp/pallets/demo/src/lib.rs:365]: why?

File: tesseract/integration-test/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (seunlanlege) [reply tesseract/integration-test/src/lib.rs:235]: just use wait_for_finalized

File: modules/ismp/pallets/demo/src/lib.rs (hunk 3 above)
Comment (MrishoLukamba) [reply modules/ismp/pallets/demo/src/lib.rs:365]: So as not to track the event and just fetch the value

Comment (seunlanlege) [reply modules/ismp/pallets/demo/src/lib.rs:365]: use wait_for_finalized

File: tesseract/integration-test/src/lib.rs (hunk 5 above)
Comment (MrishoLukamba) [reply tesseract/integration-test/src/lib.rs:474]: am doing this to make sure I wait until the chain starts to finalize blocks

File: tesseract/integration-test/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (MrishoLukamba) [reply tesseract/integration-test/src/lib.rs:235]: wait_for_event_chain_b
this is neccessary as I have to wait for event in chain b ( minting and post_request_handled),
I cannot get this via wait for finalized

**Comment by seunlanlege** [comment]
resolve conflicts so we can merge

File: tesseract/integration-test/src/lib.rs [hunk 7]
```diff
@@ -459,19 +428,40 @@ async fn get_request_works() -> Result<(), anyhow::Error> {
...
+			Ok(_block) => {
+				log::info!("chain producing blocks and finalizing");
+				break
+			},
+			Err(_err) => {
+				log::info!("chain not yet producing blocks and finalizing");
+				continue
+			},
+		}
+	}
```
Comment (seunlanlege) [inline tesseract/integration-test/src/lib.rs:448]:
```suggestion
    client_a.blocks().subscribe_finalized().await?.take(1).await;
```

**Review by seunlanlege — APPROVED** [review]

**Pull request merged**